src/dsp_tools/analyse_xml_data/extract_links_from_XML.py

```diff
@@ -0,0 +1,169 @@
+            if searched:
+                all_links.add(searched.group(1))
+    return all_links
+
+
+def _make_graph(
+    resptr_instances: list[ResptrLink], xml_instances: list[XMLLink], all_link_ids: set[str]
+) -> rx.PyDiGraph:  # type: ignore[type-arg] # pylint: disable=no-member
+    g: rx.PyDiGraph = rx.PyDiGraph()  # type: ignore[type-arg] # pylint: disable=no-member
+    nodes = [(id_, None, None) for id_ in all_link_ids]
+    node_ids = [x[0] for x in nodes]
+    node_inidices = g.add_nodes_from(nodes)
+    lookup = dict(zip(node_ids, node_inidices))
+    print(f"number of nodes: {len(nodes)}")
+    resptr_edges = [(lookup[x.subject_id], lookup[x.object_id], 1) for x in resptr_instances]
+    g.add_edges_from(resptr_edges)
+    print(f"number of resptr edges: {len(resptr_edges)}")
+    xml_edges = []
+    for xml in xml_instances:
+        xml_edges.extend([(lookup[xml.subject_id], lookup[x], xml.cost_links) for x in xml.object_link_ids])
+    g.add_edges_from(xml_edges)
+    print(f"number of xml edges: {len(xml_edges)}")
+    return g
+
+
+def _remove_leaf_nodes(g: rx.PyDiGraph) -> list[UploadResource]:  # type: ignore[type-arg] # pylint: disable=no-member
+    res: list[UploadResource] = []
+    while leaf_nodes := [x for x in g.node_indexes() if g.out_degree(x) == 0]:
+        print(f"number of leaf nodes removed: {len(leaf_nodes)}")
+        res.extend(UploadResource(g[n][0]) for n in leaf_nodes)
+        g.remove_nodes_from(leaf_nodes)
+    return res
+
+
+def _find_cheapest_node(
+    g: rx.PyDiGraph, cycle: rx.EdgeList  # type: ignore[type-arg] # pylint: disable=no-member
+) -> tuple[int, list[str]]:
+    costs = []
+    for source, _ in cycle:
+        edges_in = g.in_edges(source)
+        node_gain = len(edges_in)
+        edges_out = g.out_edges(source)
+        node_cost = sum(x[2] for x in edges_out)
+        node_value = node_cost / node_gain
+        costs.append((source, node_value, edges_out))
+    sorted_nodes = sorted(costs, key=lambda x: x[1])
+    cheapest_node, _, edges_out = sorted_nodes[0]
+    removed_target_ids = [g[x[1]][0] for x in edges_out]
+    return cheapest_node, removed_target_ids
+
+
+def _generate_upload_order(g: rx.PyDiGraph) -> list[UploadResource]:  # type: ignore[type-arg] # pylint: disable=no-member
+    removed_nodes = []
+    leaf_nodes = _remove_leaf_nodes(g)
+    removed_nodes.extend(leaf_nodes)
+    while g.num_nodes():
+        print(f"total number of nodes remaining: {g.num_nodes()}")
+        cycle = rx.digraph_find_cycle(g)  # type: ignore[attr-defined]  # pylint: disable=no-member
+        print("-" * 10)
+        print(f"cycle: {cycle}")
+        node = _find_cheapest_node(g, cycle)
+        source, targets = node
+        removed_nodes.append(UploadResource(g[source][0], targets))
+        g.remove_node(source)
+        print(f"removed link: {node}")
+        leaf_nodes = _remove_leaf_nodes(g)
+        removed_nodes.extend(leaf_nodes)
```

Comment:
@BalduinLandolt  Is it relevant that the leaf nodes are documented? Because they do not influence the size of the stash that has to be uploaded and is therefore not relevant to the end performance.

Reply:
as discussed: this is to keep track of when in the process which resource can be uploaded because it no longer has unresolvable dependencies